File: java/TinyWeb/src/com/mblinn/oo/tinyweb/TinyWeb.java
package com.mblinn.oo.tinyweb;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class TinyWeb
{

    private final List<Filter> filters;
    private final Map<String, Controller> controllers;

    public TinyWeb(final Map<String, Controller> controllers, final List<Filter> filters)
    {
        this.controllers = new HashMap<>(controllers);
        this.filters = new ArrayList<>(filters);
    }

    public HttpResponse handleRequest(final HttpRequest request)
    {
        HttpRequest currentRequest = request;
        for (final Filter filter : filters) {
            currentRequest = filter.doFilter(currentRequest);
        }

        final Controller controller = controllers.get(currentRequest.getPath());
        if (controller == null) return null;

        return controller.handleRequest(currentRequest);
    }

}

File: scala/tiny_web/src/main/scala/TinyWeb.scala
package com.mblinn.mbfpp.oo.tinyweb

import TinyWeb.Filter

class TinyWeb(val controllers: Map[String, Controller], val filters: List[Filter]) {

  lazy val composedFilter = filters reduceLeft {
    (composed, next) => composed andThen next
  }

  def handleRequest(request: HttpRequest): Option[HttpResponse] = {
    val filteredRequest = composedFilter(request)
    controllers get (filteredRequest.path) map {
      controller => controller handleRequest filteredRequest
    }
  }

}

object TinyWeb {
  type Filter = (HttpRequest) => HttpRequest
}

